SEP-2643: Structured Authorization Denials (#2643)
modelcontextprotocol/modelcontextprotocol: seps/2643-structured-authorization-denials.md

# SEP-2643: Structured Authorization Denials

- **Status**: Draft
- **Type**: Standards Track
- **Created**: 2026-04-20
- **Author(s)**: Monmohan Singh (@monmohan)
- **Sponsor**: None (Will expect Nate Barbettini - WG Lead to update)
- **PR**: https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2643

## Abstract

> This SEP defines a transport-agnostic JSON-RPC authorization denial envelope for the Model Context Protocol. The envelope complements transport-level authorization challenges (HTTP `WWW-Authenticate` with Protected Resource Metadata), carrying failure classification, a retry correlation handle, and an extensible set of structured remediation hints for cases the transport cannot easily express. This SEP defines two initial remediation hint types and illustrates their use through two scenarios: URL-based approval composed with [MCP URL-mode elicitation](https://modelcontextprotocol.io/specification/2025-11-25/client/elicitation), and credential replacement using scopes and [OAuth Rich Authorization Requests (RFC 9396)](https://datatracker.ietf.org/doc/html/rfc9396). Further remediation hint types can be defined in follow-on SEPs without changes to the envelope. This SEP is fully backward compatible. Existing OAuth clients and libraries do not need to change as the envelope is additive.

Reviewer: Propose to change "credential replacement" to "new credential issuance".
MCP Server is a resource server. It has no knowledge if the new credential will replace the existing one or be an additional one.

2. **Support for signaling server-side state remediation**: Some authorization denials are not about the client's credential. The MCP server may deny a request because it has insufficient information to contact an external system, needs confirmation from the user, or another piece of server-side state must be established before the request can succeed. OAuth authorization challenges such as `insufficient_scope` or `invalid_token` are shaped around credential changes and do not cover this case. MCP has URL-mode elicitation as a primitive for out-of-band user interaction, but it is not itself an authorization denial signaling mechanism.

> 3. **Support for structured remediation data at denial**: OAuth 2.0 defines Rich Authorization Requests (RFC 9396) for conveying structured authorization requirements, but only as part of the authorization request flow. Adopted OAuth standards do not yet provide a way to carry such structured requirements back to the client at denial time. An IETF individual draft, [`draft-zehavi-oauth-rar-metadata`](https://datatracker.ietf.org/doc/draft-zehavi-oauth-rar-metadata/02/), proposes a mechanism for HTTP by defining a new `WWW-Authenticate` error code, `insufficient_authorization_details`, with the structured data placed in a JSON response body. The draft is HTTP-specific, and this SEP proposes to adopt the same pattern at the JSON-RPC layer so it applies across MCP transports.

Reviewer: ```suggestion
3. **Support for structured remediation data at denial**: OAuth 2.0 defines Rich Authorization Requests (RFC 9396) for conveying structured authorization requirements, but only as part of the authorization request flow. Adopted OAuth standards do not yet provide a way to carry such structured requirements back to the client at denial time. An IETF individual draft, [`draft-zehavi-oauth-rar-metadata`](https://datatracker.ietf.org/doc/draft-zehavi-oauth-rar-metadata/05/), proposes a mechanism for HTTP by defining a new `WWW-Authenticate` error code, `insufficient_authorization`, with the structured data placed in the authorization_remediation header error parameter, which contains a base64url-encoded JSON object. The draft is HTTP-specific, and this SEP proposes to adopt the same pattern at the JSON-RPC layer so it applies across MCP transports.
```

Example denial:

```json
{
  "jsonrpc": "2.0",
  "id": 17,
  "error": {
    "code": -32042,
    "message": "You need to approve access to the requested folder.",
    "data": {
      "authorization": {
        "reason": "insufficient_authorization",
        "authorizationContextId": "authzctx_7c5d1d79",
        "remediationHints": [{ "type": "url" }]
      },
      "elicitations": [
        {
          "mode": "url",
          "elicitationId": "el_550e8400-e29b-41d4-a716-446655440000",
          "url": "https://mcp.example.com/approve-folder-access?ctx=authzctx_7c5d1d79",
          "message": "Open this page to approve access to the requested folder."
        }
      ]
    }
  }
}
```

After the user completes the approval, the client retries the original request, echoing the `authorizationContextId` per "Retry Echo via `_meta`":

```json
{
  "jsonrpc": "2.0",
  "id": 18,
  "method": "tools/call",
  "params": {
    "name": "files.list_folder",
    "arguments": { "folderId": "fld_12345" },
    "_meta": {
      "io.modelcontextprotocol/authorization-context-id": "authzctx_7c5d1d79"
    }
  }
}
```

> ### Use Case 2 — Credential replacement with broader authorization

Reviewer: ```suggestion
### Use Case 2 — Issuing of new credential with broader authorization
```

Reviewer: Neither client nor server know if the new credential replaces the existing one or coexists with it. But certainly UC2 result is a new credential issued

> This use case covers denials where the client's existing credential carries insufficient authorization for the requested operation and remediation requires obtaining a new credential. The replacement may involve additional OAuth scopes, a Rich Authorization Requests `authorization_details` object (RFC 9396), or any other mechanism that yields a new credential. In contrast with Use Case 1, the failure cannot be resolved by server-side state changes alone.

Reviewer: ```suggestion
This use case covers denials where the client's existing credential carries insufficient authorization for the requested operation and remediation requires obtaining a new credential, which may include additional OAuth scopes, a Rich Authorization Request's `authorization_details` object (RFC 9396), or any other mechanism to add required authorization to the new credential. In contrast with Use Case 1, the failure cannot be resolved by server-side state changes alone.
```

Author: Applied with minor changes as it was outdated

    scope="payments:initiate"
Content-Type: application/json
Cache-Control: no-store

{
  "jsonrpc": "2.0",
  "id": 31,
  "error": {
    "code": "<AUTHORIZATION_DENIAL_TBD>",
    "message": "Current authorization is insufficient for this payment.",
    "data": {
      "authorization": {
        "reason": "insufficient_authorization",
        "authorizationContextId": "authzctx_pay_9f2c",
        "remediationHints": [
          {
            "type": "oauth_authorization_details",
            "authorization_details": [
              {
                "type": "payment_initiation",
                "actions": ["initiate", "status", "cancel"],
                "locations": ["https://mcp.example.com/payments"],
                "instructedAmount": {
                  "currency": "EUR",
                  "amount": "123.50"
                },
                "creditorName": "Merchant A",
                "creditorAccount": {
                  "iban": "DE02100100109307118603"
                },
                "remittanceInformationUnstructured": "Ref Number Merchant"
              }
            ]
          }
        ]
      }
    }
  }
}
```

> The client uses the `authorization_details` from the remediation hint to construct an OAuth authorization request, per [RFC 9396#section-2](https://datatracker.ietf.org/doc/html/rfc9396#section-2):

Reviewer: ```suggestion
The client attempts to locate an existing valid credential matching a provided `authorization_reference`, if provided. If no matching credential is found, the client uses the `authorization_details` from WWW-Authenticate header's `authorization_remediation` parameter, to construct an OAuth authorization request, per [RFC 9396#section-2](https://datatracker.ietf.org/doc/html/rfc9396#section-2):
```

    }
  }
}
```

## Rationale

This SEP standardizes the smallest set of facts that an MCP client and server need to communicate after an authorization denial: what classification the failure has, how the retry is correlated, and what remediation posture the server can describe. Any further extensions, whether a new remediation type or a new transport binding, can be added additively for future needs without changing the envelope itself. For example, a future `remediationHints` type could describe a Client Initiated Backchannel Authentication (CIBA) flow, in which per-operation approvals occur on a separate channel such as a push notification rather than interrupting the foreground session.

The SEP composes with existing MCP primitives rather than inventing parallel shapes. Use Case 1 reuses MCP URL elicitation (`URLElicitationRequiredError`) and adds the envelope as sibling metadata inside the same error response. Clients that already understand URL elicitation continue to work unchanged, and clients that additionally understand the envelope receive transport-agnostic classification and a correlation handle.

> ### Future Extensions: Multi-Token Management

Reviewer: I'd remove this section completely.
UC2 with RAR relies on the rar-metadata draft, which now (version 03) includes the authorization_hint attribute, and was accepted at the OSW in Leipzig on 05/2026.
Therefore I've added to the SEP `authorization_hint` and its usage which is a one-liner actually.
Therefore I believe this comment can be removed.

Author: Agreed. Removing this para

### Client-side modification of `authorization_details`

A client MAY use the denial's `authorization_details` verbatim in a subsequent OAuth authorization request. The verbatim pattern is provided as a convenience.

Servers SHOULD structure `authorization_details` so that their contents are safe for the client to see.

Both the authorization server and the resource server are responsible for validating `authorization_details`. The authorization server validates against user consent and policy before issuing a token, and the resource server validates against the operation being requested on retry.

## Reference Implementation

TBD

## Open Questions

1. **JSON-RPC error code value**: The JSON-RPC error code introduced by this SEP is currently written as `<AUTHORIZATION_DENIAL_TBD>`. An integer value within the MCP error code range needs to be assigned before acceptance.

## Acknowledgments

This proposal was developed through discussion in the MCP Fine-Grained Authorization Working Group and refined across review cycles before being presented as a SEP, with input from:

- Nate Barbettini (MCP FGA Working Group lead)

> - Yaron Zehavi (OAuth RAR/Payment Industry expertise; author of [OAuth 2.0 RAR Metadata and Error Signaling](https://datatracker.ietf.org/doc/draft-zehavi-oauth-rar-metadata/02/))

Reviewer: ```suggestion
- Yaron Zehavi (OAuth RAR/Payment Industry expertise; author of [OAuth 2.0 RAR Metadata and Error Signaling](https://datatracker.ietf.org/doc/draft-zehavi-oauth-rar-metadata/03/))
```

```json
{
  "jsonrpc": "2.0",
  "id": 18,
  "method": "tools/call",
  "params": {
    "name": "files.list_folder",
    "arguments": { "folderId": "fld_12345" },
    "_meta": {
      "io.modelcontextprotocol/authorization-context-id": "authzctx_7c5d1d79"
    }
  }
}
```

The `authorizationContextId` is a correlation handle. Servers MAY use it for state lookup (for example, a URL-elicitation outcome in Use Case 1) and for audit and telemetry. To preserve backward compatibility with clients that do not implement this SEP, servers MUST NOT reject a retry whose handle is absent or unresolvable. Authorization is always determined by the credential presented with the request.

### Use Case 1 — Server-side state remediation via URL approval

This use case covers denials where the client's credential is valid and does not need to change. Remediation requires an out-of-band user interaction at a URL, such as approving access or selecting resources, that changes server-side state. A representative example is a file picker in a cloud storage service, where the user's OAuth token is unchanged and the server records an approval tied to the user.

> This use case composes with the MCP URL elicitation error (`URLElicitationRequiredError`, JSON-RPC error code `-32042`). The envelope MUST include a `remediationHints` entry of type `url`, and the URL elicitation remains in `data.elicitations` for backwards compatibility.

Reviewer 2: > and the URL elicitation remains in `data.elicitations` for backwards compatibility

Does this imply that it's in a different place going forward? Or is `data.elicitations` the canonical place a client should read this data from?

Author: You're right, the wording is off. data.elicitations is canonical and it isn't moving anywhere. "For backwards compatibility" isn't the real reason it's there. The real reason is that we didn't want to reinvent URL elicitation and wanted to compose, so this UC reuses -32042 as-is, and the envelope sits alongside elicitations under data as additive metadata.

That's also why the URL hint is type-only, with no pointer into the elicitation. data.elicitations is already the array clients parse according to the URL elicitation spec. A pointer would simply restate that relationship and create a risk that the two could drift.

The hint is there as a type tag so clients can continue parsing remediationHints[] uniformly across use cases, which was one of the things we wanted to achieve.
So I will update this text to reflect the choice

Author: Updated

    "_meta": {
      "io.modelcontextprotocol/authorization-context-id": "authzctx_7c5d1d79"
    }
  }
}
```

### Use Case 2 — New credential issuance with broader authorization

This use case covers denials where the client's existing credential carries insufficient authorization for the requested operation and remediation requires obtaining a new credential. This new credential issuance may involve additional OAuth scopes, a Rich Authorization Requests `authorization_details` object (RFC 9396), or any other mechanism that yields a new credential. In contrast with Use Case 1, the failure cannot be resolved by server-side state changes alone.

The envelope and the transport challenge are complementary signals:

- The envelope always carries the `reason` classification, and additionally carries the `authorizationContextId` when the server issues one.
- The envelope MAY additionally carry `remediationHints` describing structured remediation data.
- When the transport declares its own remediation (for example, an HTTP `WWW-Authenticate` challenge), the transport-level signal is authoritative for driving reauthorization. The envelope is complementary metadata.

> In UC2 on HTTP, this means: when the `WWW-Authenticate` challenge fully describes the remediation (Example 1, `insufficient_scope`), no `remediationHints` are included. When the challenge cannot fully describe the required authorization (Example 2, RAR), the envelope carries an `oauth_authorization_details` hint.

Reviewer: ```suggestion
In UC2 on HTTP, this means: when the `WWW-Authenticate` challenge fully describes the remediation (Example 1, `insufficient_scope` and example 2, `insufficient_authorization`), no `remediationHints` are included. When the challenge cannot fully describe the required authorization, the envelope carries a corresponding hint.
```

> For the stdio transport, where the MCP specification defines no transport-level authorization challenge, the envelope is the sole authorization denial signal. The client or its hosting environment is responsible for acquiring or refreshing credentials out of band before the request is retried.

Reviewer: Does this mean WWW-Authenticate header signals are not supported over stdio?

Author: Correct. As defined by the [MCP specification](https://modelcontextprotocol.io/specification/draft/basic/transports/stdio), the stdio transport carries all information within the JSON-RPC message body. The Request Metadata section explicitly states that request metadata is included inline in the body and that there is no separate header layer. This SEP follows that model.
Defining a mapping of HTTP authentication semantics, such as 401/403 responses and `WWW-Authenticate` headers, onto stdio is not something we want to pursue.

    resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource"
Content-Type: application/json

{
  "jsonrpc": "2.0",
  "id": 17,
  "error": {
    "code": "<AUTHORIZATION_DENIAL_TBD>",
    "message": "You must grant write access to proceed.",
    "data": {
      "authorization": {
        "reason": "insufficient_authorization",
        "authorizationContextId": "authzctx_8a3e1f4b"
      }
    }
  }
}
```

The client performs OAuth reauthorization using the transport challenge, obtains a new access token whose scope includes both `read` and `write`, and retries the original request with the new token in the `Authorization` header. The client echoes the `authorizationContextId` per "Retry Echo via `_meta`":

```http
POST /mcp HTTP/1.1
Host: mcp.example.com
Authorization: Bearer at_new
Content-Type: application/json

{
  "jsonrpc": "2.0",
  "id": 18,
  "method": "tools/call",
  "params": {
    "name": "files.update",
    "arguments": { "fileId": "fl_98765", "content": "..." },
    "_meta": {
      "io.modelcontextprotocol/authorization-context-id": "authzctx_8a3e1f4b"
    }
  }
}
```

#### Example 2 — Rich Authorization Requests remediation

> When a denied request requires authorization beyond what OAuth scopes can express, the `WWW-Authenticate` challenge can signal that authorization is insufficient but cannot carry the structured requirements themselves. The server MAY supply those requirements inside the envelope so the client can construct its next authorization request directly, without additional metadata discovery. This approach mirrors the error-signaling pattern defined in [OAuth 2.0 RAR Metadata and Error Signaling](https://datatracker.ietf.org/doc/draft-zehavi-oauth-rar-metadata/02/) at the JSON-RPC layer.

Reviewer: ```suggestion
When a denied request requires authorization beyond what OAuth scopes can express, the `WWW-Authenticate` challenge can signal that authorization is insufficient and also provide the structured remediation requirements, with which the client can construct its next authorization request directly, without additional metadata discovery. This approach uses the error-remediation pattern defined in [OAuth 2.0 RAR Metadata and Error Signaling](https://datatracker.ietf.org/doc/draft-zehavi-oauth-rar-metadata/05/).
```

> An MCP server that denies a request because the access token lacks sufficient authorization details MAY return a `remediationHints` entry of type `oauth_authorization_details`. The entry's `authorization_details` member, when present, SHALL be an OAuth Rich Authorization Requests `authorization_details` array as defined in RFC 9396. On HTTP Transports, The server MUST still return an HTTP authorization challenge via `WWW-Authenticate`, which remains authoritative for remediation and discovery. The JSON-RPC remediation hint is supplemental and exists to carry an actionable `authorization_details` object inside the response body.

Reviewer: ```suggestion
An MCP server that denies a request because the access token lacks sufficient authorization details returns a transport layer WWW-Authenticate header with appropriate remediation requirements. The client performs OAuth reauthorization using the transport challenge, obtains a new access token which includes the provided `authorization_details`, and retries the original request with the new token in the `Authorization` header. The client echoes the `authorizationContextId` per "Retry Echo via `_meta`".
```

> A client that recognizes the `oauth_authorization_details` hint MAY use the provided `authorization_details` directly in a subsequent OAuth authorization request. A client that does not recognize the hint SHOULD fall back to resource-metadata and authorization-server-metadata discovery to determine how to construct a valid authorization request.

Reviewer: ```suggestion
A client that recognizes the `WWW-Authenticate` header with `insufficient_authorization` error code and `authorization_remediation` parameter MAY use an existing valid credential matching a provided `authorization_reference` value, if provided. If no matching credential is found, client MAY use the provided `authorization_details` directly in a subsequent OAuth authorization request. A client that does not recognize the hint SHOULD fall back to resource-metadata and authorization-server-metadata discovery to determine how to construct a valid authorization request.
```

Consider an MCP server that exposes a `payments.initiate` tool where initiating a payment requires authorization details bound to the payee and amount. The client's current access token does not carry such authorization details, and the call is denied:

```json

> HTTP/1.1 403 Forbidden

Reviewer: ```suggestion
HTTP/1.1 401 Unauthorized
```

> WWW-Authenticate: Bearer error="insufficient_authorization_details",

Reviewer: ```suggestion
WWW-Authenticate: Bearer error="insufficient_authorization",
    resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource/payments",
    error_description="Additional authorization is required",
    authorization_remediation=eyJhdXRob3JpemF0aW9uX2RldGFpbHMiOlt7InR5cGUiOiJkaX...
```

>     resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource/payments",

Reviewer: Remove, included above

    scope="payments:initiate"
Content-Type: application/json
Cache-Control: no-store

{
  "jsonrpc": "2.0",
  "id": 31,
  "error": {
    "code": "<AUTHORIZATION_DENIAL_TBD>",
    "message": "Current authorization is insufficient for this payment.",
    "data": {
      "authorization": {
        "reason": "insufficient_authorization",
        "authorizationContextId": "authzctx_pay_9f2c",

>         "remediationHints": [

Reviewer: Propose remove `remediationHints`, it's provided in the header

Author: Removed. The challenge carries the remediation in this example, so the envelope carries classification only.

A client that has obtained a remediation, whether from the transport challenge or from an `authorization_remediation` hint, MAY first check whether it already holds a valid credential matching the `authorization_reference`, when one is provided, and retry using that credential. Otherwise the client MAY use the provided `authorization_details` directly in a subsequent OAuth authorization request. On HTTP transports, a client that does not recognize the remediation SHOULD fall back to resource-metadata and authorization-server-metadata discovery to determine how to construct a valid authorization request.

Consider an MCP server that exposes a `payments.initiate` tool where initiating a payment requires authorization details bound to the payee and amount. The client's current access token does not carry such authorization details, and the call is denied:

```json
HTTP/1.1 401 Unauthorized
WWW-Authenticate: Bearer error="insufficient_authorization",

>     resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource/payments",

Reviewer: I think we can remove the resource_metadata url

Author: Hi Yaron, This is following Section 7.1 from your draft -

> If the authorization server used so far by the client does not support the required authorization details types, the client MAY use Protected Resource Metadata [[RFC9728](https://www.rfc-editor.org/rfc/rfc9728)] to discover additional authorization_servers supported by the resource, and attempt remediation through them.

Let me know if you would still recommend to remove it

Reviewer: right, keep it

## Abstract

This SEP defines a transport-agnostic JSON-RPC authorization denial envelope for the Model Context Protocol. The envelope complements transport-level authorization challenges (HTTP `WWW-Authenticate` with Protected Resource Metadata), carrying failure classification, a retry correlation handle, and an extensible set of structured remediation hints for cases the transport cannot easily express. This SEP defines two initial remediation hint types and illustrates their use through three scenarios: URL-based approval composed with [MCP URL-mode elicitation](https://modelcontextprotocol.io/specification/2025-11-25/client/elicitation), credential replacement using scopes and [OAuth Rich Authorization Requests (RFC 9396)](https://datatracker.ietf.org/doc/html/rfc9396), and additional short-lived credentials for per-transaction operations like payment initiation in banking. Further remediation hint types can be defined in follow-on SEPs without changes to the envelope. This SEP is fully backward compatible. Existing OAuth clients and libraries do not need to change as the envelope is additive.

## Motivation

> 1. **Stdio has no defined authorization challenge mechanism**: The MCP specification defines authorization procedures for the HTTP transport, including OAuth 2.1 flows, `WWW-Authenticate` challenges, and Protected Resource Metadata discovery. These mechanisms explicitly exclude stdio, which the specification directs to rely on credentials retrieved from the environment instead. As a result, a stdio MCP server that needs to reject a request on authorization grounds has no standard mechanism to communicate that the failure is an authorization problem, to indicate the nature of the required remediation, or to carry structured data the client can act on.

Reviewer 3: Additionally, current guidance is that STDIO servers that require OAuth remediation steps must handle the OAuth flow themselves and act as the OAuth client. This obscures the identity of the MCP Client from other participants in the system.

Author: Agreed. Will add this to Motivation#1

Reviewer 2: I agree, although I do want to make sure we don't venture into trying to solve auth for stdio.

> 2. **Support for signaling server-side state remediation**: Some authorization denials are not about the client's credential. The server may deny a request because user approval, resource selection, or another server-side state must be established through direct user interaction. OAuth authorization challenges such as `insufficient_scope` or `invalid_token` are shaped around credential changes and do not cover this case. MCP has URL-mode elicitation as a primitive for out-of-band user interaction, but it is not itself an authorization denial signaling mechanism.

Reviewer 3: I want to echo Yaron's point below about conflating of OAuth roles - as this reads, it indicates that the MCP Server sometimes requires explicit user consent. However, the collection and management of user consent is in the domain of the AS. Suggest leading with a different example which is not as clearly within the domain of the AS.

Perhaps `insufficient information to contact upstream systems`, e.g. needing to collect credit card details?

Author: Fair enough. I propose changing

**Current wording:**
Some authorization denials are not about the client's credential. The server may deny a request because user approval, resource selection, or another server-side state must be established through direct user interaction.

**To Updated Wording:**
Some authorization denials are not about the client's credential. The server may deny a request because information needed to contact upstream systems (for example, payment details collected from the user), resource selection, or another piece of server-side state must be established through direct user interaction.

Does that address the concern?

Reviewer 2: In general I would suggest being very explicit about `MCP server` in this doc rather than just saying `server`. Perhaps a further updated wording could be:

> Some authorization denials are not about the client's credential. The MCP server may deny a request because it has insufficient information to contact an external system, needs confirmation from the user, or another piece of server-side state must be established before the request can succeed.

Author: Agreed on staying in lane. The two sentences I appended describe the status quo and its downside, not a stdio auth solution.

4. **Support for additive credential flows**: OAuth 2.0 defines token lifetime through the `expires_in` parameter but does not define the lifecycle relationship between a newly-issued credential and any existing credential the client already holds. General-purpose clients therefore typically default to replacing the bearer token on new issuance, which does not support use cases where a short-lived credential for a specific operation is issued in addition to a long-lived credential held by the client. An example is a banking client that holds a long-lived account-information token and must obtain a separate short-lived per-payment authorization for each transaction, while continuing to use the long-lived token for other operations.

## Specification

### Authorization Denial Envelope

> An MCP server that denies a JSON-RPC request on authorization grounds MAY return a JSON-RPC error whose `code` is `<AUTHORIZATION_DENIAL_TBD>` (integer value to be assigned, see Open Questions) and whose `data` contains an `authorization` object populated as defined in this section. The envelope is transport-agnostic and applies to both HTTP and stdio transports.

Reviewer 2: There may be more transports than just 2 in the future!
```suggestion
An MCP server that denies a JSON-RPC request on authorization grounds MAY return a JSON-RPC error whose `code` is `<AUTHORIZATION_DENIAL_TBD>` (integer value to be assigned, see Open Questions) and whose `data` contains an `authorization` object populated as defined in this section. The envelope is transport-agnostic.
```

> The conceptual shape is:

Reviewer 2: ```suggestion
The shape of the payload is:
```

  "id": 21,
  "error": {
    "code": "<AUTHORIZATION_DENIAL_TBD>",
    "message": "Authorization is required to continue.",
    "data": {
      "authorization": {
        "reason": "insufficient_authorization",
        "authorizationContextId": "authzctx_6f2b0d3e",
        "credentialDisposition": "<replacement | additional>",
        "remediationHints": [
          {
            "type": "<remediation-type>"
            /* additional members determined by the remediation type */
          }
        ]
      }
    }
  }
}
```

The fields of the `authorization` object are defined as follows:

- `reason` (string, REQUIRED). Indicates that the failure is an authorization denial and classifies its general nature. This SEP defines a single value, `insufficient_authorization`. Additional values MAY be defined by future SEPs.
- `authorizationContextId` (string, OPTIONAL). A server-issued correlation handle. When the server includes this field, the client MUST echo it on retry as described in "Retry Echo via `_meta`". The handle is not authorization material, and the server MUST NOT rely on it for authorization decisions when the credential presented on retry is sufficient on its own.

> - `credentialDisposition` (string, OPTIONAL). Describes the lifecycle relationship between a credential newly issued through the remediation flow and any existing credential held by the client. Defined values:

Reviewer: I struggle with this part as it conflates classic OAuth roles, whereby the lifecycle of tokens it for the authorization server (AS) to say, and not in the resource server's (RS) domain.
This introduces undesirable coupling between AS and RS.
The MCP Server might accept tokens from several authorization servers, to properly serve this response attribute it would need to maintain metadata pertaining to how each of the authorization servers handles the remediation.
Furthermore this metadata even if managed, may change over time.
AS and RS might be from different providers, and there are no established mechanisms to communicate this added metadata.

I see it simpler, remove credentialDisposition and:

- Client should decide whether it wants to request a **separate**, **limited-use** credential for remediation purpose only, or make a request "**upgrading**" to a multi-purpose credential, combining the remediation RAR and **scopes previously granted**.
- Client should then assess credential response's **expires_in** attribute, to decide if previously existing creedntials should be retained.
- Note - OAuth has no normative "single-use semantics" for access tokens. Each grant is independent and bears no normative outcome on results of previously completed grants.

Reviewer 3: Agreed that the tight coupling of token semantics to resource is not great.

> Client should then assess credential response's expires_in attribute, to decide if previously existing creedntials should be retained.

This works with scope semantics where access tokens can be compared against each other directly, but not with RAR semantics. If clients are dealing with many resources that require RAR semantics, they may end up dealing with many tokens that exist at the same time. We need to define a strategy for how clients can manage multiple tokens and select the appropriate one to use for each request.

Reviewer: The rar metadata draft offers resource server rar semantics to help clients navigate which token satisfies which resource:
https://www.ietf.org/archive/id/draft-zehavi-oauth-rar-metadata-02.html#name-oauth-20-protected-resource

According to the proposal, RFC 9728 resource server metadata response shall include `authorization_details_types_supported` which guides clients on choosing the right token (compare required RAR types to those granted to tokens they have).

Also worth mentioning that the challenge is simplified in case short lived tokens are used by authorization servers for high-risk RAR use cases, as clients will prune expired tokens.

Reviewer 3: `authorization_details_types_supported` is a good start, but how can clients make this determination when clients have multiple tokens with the same `type`?

For example, say that a client is involved in two transactions simultaneously. THe client has one token containing the following details object: 
```
{
  "type": "payment_initiation",
  "instructedAmount": { "currency": "EUR", "amount": "100.00" },
  "creditorAccount": { "iban": "AAAAAAAAAAA"}
}
```

And another token containing the following details object:
```
{
  "type": "payment_initiation",
  "instructedAmount": { "currency": "GBR", "amount": "50.00" },
  "creditorAccount": { "iban": "BBBBBBBBBBBB"}
}
```

How can the client choose which token to send for each request in a deterministic and efficient manner?

`authorization_details_types` describes a **class** of token, but the client needs to determine the correct **instance** of token.

Author: Thanks @yaron-zehavi and @max-stytch — both points well taken. What you are collectively pointing at is that MCP clients MAY need a deterministic way to manage multiple credentials, and UC3 as currently described in this SEP, especially the `credentialDisposition: additional` framing tied to client-side single-use semantics, is not the right design for that. I see @max-stytch has posted a doc on multi-token management for RAR to start the discussion. So I think once there is convergence on the problem and design then we can decide what UC3 should look like in this SEP and update accordingly

Reviewer: The challenge of client selecting the right token among several options is generic and not specific to MCP.
And I think the challenges it involves, including @max-stytch 's example are navigable without requiring new normative language and without requiring clients to understand RAR object structure:

Clients of Authorization Servers that issue discrete separate short lived RAR tokens, simplify client token selection, because operations permitted by said tokens are handled one at a time in a short time frame, and said RAR tokens quickly expire: client doesn't end up with 2 payment tokens in parallel.

Clients of Authorization Servers that issue long lived RAR tokens, e.g to represent a permission policy, can: 
1. Request to incrementally add additionally required RAR objects and Scopes to a single credential -> resulting in one token, no confusion what to choose.
2. If client has multiple credentials, choose based on RAR type satisfying resource's requirements.
3. If duplicates exist, try using all existing tokens.
4. If none work, request new token

Reviewer 2: @yaron-zehavi I think I follow, but I still struggle to see what guidance we must give to MCP client developers.

Stepping back, if we introduce the idea in this SEP that the client MAY have to keep track of multiple tokens, then we MUST provide deterministic guidance on how to do it. (Notably the baseline today is that MCP clients are not expected to handle more than 1 access token.)

Reviewer 2: I see @max-stytch is already ahead of me in digging into that problem 🙂 thanks!

Author: UC3 is removed from the SEP, along with the related Motivation #4, the `credentialDisposition` envelope field, and the UC3-specific Security Implications subsection.
There is a new Rationale subsection, “Future Extensions: Multi-Token Management,” that names the deferred problem, points to the planned extension of [draft-zehavi-oauth-rar-metadata](https://datatracker.ietf.org/doc/draft-zehavi-oauth-rar-metadata/02/), and acknowledges Max’s and the WG’s work on client-side multi-credential handling.

  - `replacement` (default). The newly-issued credential supersedes any existing credential for the same grant context. This matches current OAuth behavior and is the implicit default when `credentialDisposition` is absent.

>   - `additional`. The newly-issued credential coexists with the existing credential. The client MUST retain the existing credential for its original purpose. The newly-issued credential's usable scope is determined by its grant, and the client selects the appropriate credential per request.

Reviewer 3: > the client selects the appropriate credential per request.

We need a well-defined strategy for this. I will try to jot down some notes and examples later, but I wonder if the client can 
- Try a request with its "base credential"
- Look at the WWW-Authenticate response
- Attempt to match the challenge to a "specific credential" within its bag of tokens
- Retry with the specific credential
- Attempt to get a new credential only on failure

Reviewer: I'd add that in step 1 client can consult (rar-metadata draft's proposed edition to) resource's metadata endpoint for `authorization_details_types_supported` to guide choosing a satisfying credential.
Cllients that did so can skip to step 5 and request a new credential using provided `remediationHints`.

Reviewer 3: Yes, short-circuiting is definitely possible.

For those not in the discord- We started discussing credential selection strategies / schemes in https://docs.google.com/document/d/1X4q8qVazt2MIuLs2kMpAqJn5rlyYnBGk0FLw2a9D-yU/edit?tab=t.0#heading=h.fpw5paiaqpja

Reviewer: As this is a challenge faced successfully by clients outside MCP, I'm not sure a new normative language is required.
However, certainly outlining strategies using current and upcoming metadata is helpful.

Author: See https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2643#discussion_r3292084982

>   The lifetime of an additional credential is governed by its OAuth `expires_in` value. Client-side management of additional credentials (storage, selection per request, cleanup after expiry) is an implementation concern outside the scope of this SEP. Servers that require single-use semantics for an additional credential SHOULD enforce single-use at the resource server.

Reviewer 3: I don't think single use semantics should be modeled as an inherent property of the credential that need to be enforced. Instead, they are a property of the resource that the credential has the ability to act on. For example:
1. A client wants to transfer money from account A to account B, and tries to do so with params `{ from, to, amount }`
2. The server binds the params to an opaque `transfer_id` and returns that `transfer_id` within the RAR payload
3. The AS authorizes the client with a token scoped to that `transfer_id`
4. The client retries the original request with the token, which succeeds

The single-use semantics are a property of the Transfer resource itself - once the transaction is complete the access token is useless because it has no more resources to operate on, not because it has been used.

This means that an access token can be used across:
- Multiple related tools that operate on the same transfer (or other short-lived object)
- Multiple tool calls that fail due to network issues, intermittent issues, etc.

Reviewer: I think this makes assumptions that don't always hold true:

1. Server's might avoid using a unique operation_id inside RAR object and instead enforce single-use by caching a used access token until its expiry
2. Some servers may allow using said access token on several tools, some only on one

Reviewer 2: Reflecting the "we must provide guidance to MCP client developers" point above, I don't think we should say here that e.g. token selection per request is a concern outside the scope of this SEP.

Author: See https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2643#discussion_r3292084982

- `remediationHints` (array of objects, OPTIONAL). Structured remediation hints carried at the JSON-RPC layer that complement any transport-level authorization challenge. Each hint is an object with:
  - a `type` field (string, REQUIRED), naming the remediation mechanism, and
  - zero or more additional members whose names and shapes are determined by the value of `type`.

  Clients MUST ignore remediation hints whose `type` they do not recognize.

> The envelope is additive. When a transport-level authorization challenge is present, for example an HTTP `WWW-Authenticate` header, that challenge remains authoritative for driving the client's authorization flow. The envelope provides a transport-agnostic failure classification and, where applicable, structured data that the transport challenge cannot easily express. For stdio, where the MCP specification defines no transport-level authorization challenge mechanism, the envelope is the sole authorization denial signal.

Reviewer 3: What happens when the envelope and the header disagree with each other? (for example, a scope mismatch)

Reviewer: See line 152, such disagreement imo is avoided:
"When the challenge fully describes the required remediation (for example `insufficient_scope` with the required scopes), the envelope functions as pure classification metadata carrying `reason` and an optional `authorizationContextId`. "

Author: Addressed by Yaron's reply. Resolving

### Retry Echo via `_meta`

> When a server includes an `authorizationContextId` in the authorization denial envelope, the client MUST echo the value, verbatim, in the `_meta` block of the retry request under the key `io.modelcontextprotocol/authorization-context-id`.

Reviewer 3: Since this is an official SEP and not an extension, do we need to use the `_meta` field?

Author: My reasoning was the following:
1. The echo is conditional. Clients MUST echo authorizationContextId only when the server issued one.
2. Didn't want to expand the core envelope on every SEP — _meta absorbs conditional fields without changing top-level shape.
3. As per my read of https://modelcontextprotocol.io/specification/2025-11-25/basic#_meta - it's for both first-party and 3rd-party. The difference is via the prefix `io.modelcontextprotocol/` namespace.

But happy to hear what others think

Example retry after a denial carrying `authzctx_7c5d1d79`:

```json
{
  "jsonrpc": "2.0",
  "id": 18,
  "method": "tools/call",
  "params": {
    "name": "files.list_folder",
    "arguments": { "folderId": "fld_12345" },
    "_meta": {
      "io.modelcontextprotocol/authorization-context-id": "authzctx_7c5d1d79"
    }
  }
}
```

> Authorization of the retry is determined by the credential presented with the request, and the server MUST NOT reject a retry solely because the handle is absent or cannot be resolved.

Reviewer 2: If the server MUST NOT reject, then it is effectively optional. In that case, what server behavior does it drive?

Author: I restructured the section The handle is used to correlate the denial with the retry. It also supports state lookup for UC1, audit, and telemetry. The `MUST NOT reject` requirement for a missing handle is there for backward compatibility. Clients that do not implement this SEP will not echo the handle back, so the server still needs to accept retries with or without it.  I would have preferred stricter enforcement so the server could rely on the handle always being present, but that would break pre-SEP clients.
The current approach keeps SEP-aware clients consistent while leaving room to tighten enforcement in a future revision.

### Use Case 1 — Server-side state remediation via URL approval

This use case covers denials where the client's credential is valid and does not need to change. Remediation requires an out-of-band user interaction at a URL, such as approving access or selecting resources, that changes server-side state. A representative example is a file picker in a cloud storage service, where the user's OAuth token is unchanged and the server records an approval tied to the user.

> This use case composes with the MCP URL elicitation error (`URLElicitationRequiredError`, JSON-RPC error code `-32042`). The envelope MUST include a `remediationHints` entry of type `url`, and the URL elicitation remains in `data.elicitations`.

Reviewer 2: ```suggestion
This use case composes with the MCP URL elicitation error (`URLElicitationRequiredError`, JSON-RPC error code `-32042`). The envelope MUST include a `remediationHints` entry of type `url`, and the URL elicitation remains in `data.elicitations` for backwards compatibility.
```

          "url": "https://mcp.example.com/approve-folder-access?ctx=authzctx_7c5d1d79",
          "message": "Open this page to approve access to the requested folder."
        }
      ]
    }
  }
}
```

After the user completes the approval, the client retries the original request, echoing the `authorizationContextId` per "Retry Echo via `_meta`":

```json
{
  "jsonrpc": "2.0",
  "id": 18,
  "method": "tools/call",
  "params": {
    "name": "files.list_folder",
    "arguments": { "folderId": "fld_12345" },
    "_meta": {
      "io.modelcontextprotocol/authorization-context-id": "authzctx_7c5d1d79"
    }
  }
}
```

### Use Case 2 — Credential replacement with broader authorization

This use case covers denials where the client's existing credential carries insufficient authorization for the requested operation and remediation requires obtaining a new credential. The replacement may involve additional OAuth scopes, a Rich Authorization Requests `authorization_details` object (RFC 9396), or any other mechanism that yields a new credential. In contrast with Use Case 1, the failure cannot be resolved by server-side state changes alone.

> On the HTTP transport, the `WWW-Authenticate` challenge remains authoritative for driving the client's reauthorization. When the challenge fully describes the required remediation (for example `insufficient_scope` with the required scopes), the envelope functions as pure classification metadata carrying `reason` and an optional `authorizationContextId`. When the required authorization cannot be fully described at the transport layer, the envelope MAY carry a structured remediation hint as described in Example 2 below.

Reviewer 2: I agree with this, but I wonder if we can make the relationship between the payload and the transport clearer here. I think the points we want to communicate are:
- The envelope always carries complementary information (reason, context ID)
- The envelope MAY carry a remediation hint
- If the transport declares a remediation, it is authoritative

Are there any nuances I missed in those bullets?

Author: This is already covered in the paragraph but I think the bullets read better so changed it accordingly.

    resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource"
Content-Type: application/json

{
  "jsonrpc": "2.0",
  "id": 17,
  "error": {
    "code": "<AUTHORIZATION_DENIAL_TBD>",
    "message": "You must grant write access to proceed.",
    "data": {
      "authorization": {
        "reason": "insufficient_authorization",
        "authorizationContextId": "authzctx_8a3e1f4b"
      }
    }
  }
}
```

The client performs OAuth reauthorization using the transport challenge, obtains a new access token whose scope includes both `read` and `write`, and retries the original request with the new token in the `Authorization` header. The client echoes the `authorizationContextId` per "Retry Echo via `_meta`":

```http
POST /mcp HTTP/1.1
Host: mcp.example.com
Authorization: Bearer at_new
Content-Type: application/json

{
  "jsonrpc": "2.0",
  "id": 18,
  "method": "tools/call",
  "params": {
    "name": "files.update",
    "arguments": { "fileId": "fl_98765", "content": "..." },
    "_meta": {
      "io.modelcontextprotocol/authorization-context-id": "authzctx_8a3e1f4b"
    }
  }
}
```

#### Example 2 — Rich Authorization Requests remediation

> When a denied request requires authorization beyond what OAuth scopes can express, the `WWW-Authenticate` challenge can signal that authorization is insufficient but cannot carry the structured requirements themselves. The server MAY supply those requirements inside the envelope so the client can construct its next authorization request directly, without navigating the metadata discovery chain. This approach mirrors the error-signaling pattern defined in [OAuth 2.0 RAR Metadata and Error Signaling](https://datatracker.ietf.org/doc/draft-zehavi-oauth-rar-metadata/02/) at the JSON-RPC layer.

Reviewer 2: ```suggestion
When a denied request requires authorization beyond what OAuth scopes can express, the `WWW-Authenticate` challenge can signal that authorization is insufficient but cannot carry the structured requirements themselves. The server MAY supply those requirements inside the envelope so the client can construct its next authorization request directly, without additional metadata discovery. This approach mirrors the error-signaling pattern defined in [OAuth 2.0 RAR Metadata and Error Signaling](https://datatracker.ietf.org/doc/draft-zehavi-oauth-rar-metadata/02/) at the JSON-RPC layer.
```

Reviewer: Perhaps "without **requiring** additional metadata discovery."

> An HTTP MCP server that denies a request because the access token lacks sufficient authorization details MAY return a `remediationHints` entry of type `oauth_authorization_details`. The entry's `authorization_details` member, when present, SHALL be an OAuth Rich Authorization Requests `authorization_details` array as defined in RFC 9396. The server MUST still return an HTTP authorization challenge via `WWW-Authenticate`, which remains authoritative for remediation and discovery. The JSON-RPC remediation hint is supplemental and exists to carry an actionable `authorization_details` object inside the response body.

Reviewer 2: I think this guidance is broader than just HTTP - any MCP server that denies a request could return `oauth_authorization_details`, right?

A client that recognizes the `oauth_authorization_details` hint MAY use the provided `authorization_details` directly in a subsequent OAuth authorization request. A client that does not recognize the hint SHOULD fall back to resource-metadata and authorization-server-metadata discovery to determine how to construct a valid authorization request.

Consider an MCP server that exposes a `payments.initiate` tool where initiating a payment requires authorization details bound to the payee and amount. The client's current access token does not carry such authorization details, and the call is denied:

```json
HTTP/1.1 403 Forbidden

> WWW-Authenticate: Bearer error="insufficient_authorization_details",

Reviewer 3: @yaron-zehavi this is a really good example of where passing `authorization_details in response body` as described in the base draft doesn't work well. Individual ecosystems rely on the response body for their own unique error messages.

Being able to pass authorization details (or a reference to the details - GNAP style) in the `WWW-auth` header ensures the response body is free for application and ecosystem specific logic.

Reviewer: I think it is the accepted convention that authentication & authorization error handling MAY trump ecosystem application standards.
So an HTTP 401 is usually going to return (following RFC 7807) something like:

HTTP 401 Unauthorized
WWW-Authenticate: Bearer error="invalid_token"
Content-Type: application/problem+json

{
  "type": "https://example.com/errors/invalid-token",
  "title": "Invalid access token",
  "status": 401,
  "detail": "The token has expired"
}

... And not something very different according to the ecosystem.

Author: In addition to what @yaron-zehavi  says, note that a JSON-RPC 2.0 error is a fixed envelope (code, message, data), and we are placing the denial under a single named key, data.authorization. Servers are free to populate other keys in data for their own information.

A couple of reasons where the body helps in our context:

1. Encoding practicality and header size. Realistic authorization_details arrays for cases like payment_initiation may run into header size limits. JSON in a body avoids that.
2. Uniform envelope mechanism across transports. For a stdio MCP server to bubble up the denial to its MCP client (the pattern you mention in your Motivation#1 comment), the JSON-RPC body is the right carrier. keeps one parsing model across both transports.

NOTE -

1. When `oauth_authorization_details` is present, a recognizing client MAY use the array directly to construct its next authorization request, saving a metadata-discovery round trip. The SEP already requires AS and RS to validate, so the hint is not a trust boundary in either direction. The hint is opt-in. A server that prefers to keep the requirements out of band — sensitivity, integrity concerns, payload size — can omit it and leave the client on the WWW-Authenticate + resource-metadata path. 
2. The SEP doesn't preclude future header-reference designs (such as the resource-set registration reference under discussion in [https://github.com/yaron-zehavi/oauth-rich-authorization-requests-metadata/issues/4](https://github.com/yaron-zehavi/oauth-rich-authorization-requests-metadata/issues/4), which I believe is what you are referring to, @max-stytch ).

Author: No changes needed. Resolving

>     resource_metadata="https://mcp.example.com/.well-known/oauth-protected-resource/payments",

Reviewer 2: Is a different `resource_metadata` URL required, or just a non-normative example here?

Reviewer: The idea is that MCP Server would likely require different RAR object types per resource.
Therefore the resource_metadata url will match the resource (e.g /payments in the example here).
resource_metadata url's response would indicate the required RAR types, e.g:
    "authorization_details_types_supported": {
        "oneOf": ["payment_initiation", "payment_approval",
                  "beneficiary_designation"]

See in my draft:
https://www.ietf.org/archive/id/draft-zehavi-oauth-rar-metadata-02.html#name-oauth-20-protected-resource

Reviewer 2: Makes sense, but are we saying here that Structured Authorization Denials _requires_ this idea, or that it is merely supported? (Normative vs. non-normative)

Author: Non-normative. SEP defers to OAuth as authoritative for HTTP transport remediation.  The per-resource pattern in the example follows the approach Yaron described above from his RAR-metadata draft.  `resource_metadata ` and its URL structure are defined by OAuth Protected Resource Metadata. So, Any URL valid under OAuth PRM is acceptable.

    scope="payments:initiate"
Content-Type: application/json
Cache-Control: no-store

{
  "jsonrpc": "2.0",
  "id": 31,
  "error": {
    "code": "<AUTHORIZATION_DENIAL_TBD>",
    "message": "Current authorization is insufficient for this payment.",
    "data": {
      "authorization": {
        "reason": "insufficient_authorization",
        "authorizationContextId": "authzctx_pay_9f2c",
        "remediationHints": [
          {
            "type": "oauth_authorization_details",
            "authorization_details": [
              {
                "type": "payment_initiation",
                "actions": ["initiate", "status", "cancel"],
                "locations": ["https://mcp.example.com/payments"],
                "instructedAmount": {
                  "currency": "EUR",
                  "amount": "123.50"
                },
                "creditorName": "Merchant A",
                "creditorAccount": {
                  "iban": "DE02100100109307118603"
                },
                "remittanceInformationUnstructured": "Ref Number Merchant"
              }
            ]
          }
        ]
      }
    }
  }
}
```

> The client uses the `authorization_details` from the remediation hint to construct an OAuth authorization request:

Reviewer 2: Can you add specific client guidance (or reference an RFC) on how to do this?

Reviewer: https://datatracker.ietf.org/doc/html/rfc9396

Author: Added a reference to [RFC 9396#section-2](https://datatracker.ietf.org/doc/html/rfc9396#section-2)

```http
GET /authorize?
    response_type=code&
    client_id=client-123&
    redirect_uri=https%3A%2F%2Fclient.example.com%2Fcb&
    scope=payments%3Aapprove&
    authorization_details=%5B%7B...payment_initiation...%7D%5D HTTP/1.1
Host: as.example.com
```

After obtaining a new access token, the client retries the original request, echoing the `authorizationContextId`:

```json
POST /mcp HTTP/1.1
Host: mcp.example.com
Authorization: Bearer at_new
Content-Type: application/json

{
  "jsonrpc": "2.0",
  "id": 32,
  "method": "tools/call",
  "params": {
    "name": "payments.initiate",
    "arguments": {
      "amount": "123.50",
      "currency": "EUR",
      "creditorName": "Merchant A",
      "creditorAccount": "DE02100100109307118603",

>       "remittanceInformation": "Ref Number Merchant"

Reviewer 2: nit: This was called `remittanceInformationUnstructured` in the `authorization_details` block

    },
    "_meta": {
      "io.modelcontextprotocol/authorization-context-id": "authzctx_pay_9f2c"
    }
  }
}
```

> ### Use Case 3 — Additional short-lived credential

Reviewer: If you accept my previous comment, UC3 can be removed completely as in general an MCP server should examine the remediating credential based on it containing the missing scope / authorization details, being not expired and not previously used.

An MCP server should in general not care about the lifetime of the credential in taking authorization decisions (except as mentioned alredy enforce that it's not expired).

Author: please see https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2643#discussion_r3181190943

Author: See https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2643#discussion_r3292084982

### Correlation handle is not authorization material

The `authorizationContextId` field is a correlation handle, not authorization material. The credential on the retry, not just the handle, is what determines whether the request is authorized.

### Single-use enforcement for `credentialDisposition: additional`

For `credentialDisposition: additional`, single-use or per-operation constraints on the additional credential are not security boundaries when expressed client-side. Servers that require them SHOULD enforce them at the resource server.

### Information disclosure through remediation hints

Remediation hints describe what a client should do in response to a denial. They MUST NOT carry information that would itself require authorization to read, such as the identity of another user whose approval would be required or the contents of a resource the client cannot access.

> ### Client-side modification of `authorization_details`

Reviewer 3: ref: https://github.com/yaron-zehavi/oauth-rich-authorization-requests-metadata/issues/4

Reviewer: I think this merits a separate issue, see:
https://github.com/yaron-zehavi/oauth-rich-authorization-requests-metadata/issues/5

Author: No changes needed. Resolving
